### DISCUSSION TITLE
Invalid HTTP request received

### DISCUSSION THREAD

**Discussion by darath2000** [Kludex/uvicorn#2487] [Potential Issue]
Hi.

I am reaching out to seek clarification on a potential issue I have encountered while working with a FastAPI project that utilizes Uvicorn. Specifically, I am unsure whether the behavior I'm observing is a bug or expected functionality.

When I execute the following code snippet:
```
import socket

host = 'localhost'
port = 8000

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
    sock.connect((host, port))
    request = b'GET /\r\n\r\n'
    sock.sendall(request)
    response = sock.recv(4096)
    print(response.decode())
```
I receive the following warning on the server:
**Invalid HTTP request received.**
Given that b'GET /\r\n\r\n' appears to be a valid HTTP request, I would like to understand if this behavior is expected or if there might be an underlying issue in uvicorn.

Server code:
```
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def hello_world():
    return {"message": "Hello World"}
```

**Comment by vvanglro** [comment]
This problem has nothing to do with uvicorn, this is what h11 returns when parsing the message.

h11 and httptools behave differently when parsing this message.
h11 parsing error, httptools normal parsing, they all seem to have their own standards.
```py
import h11

request = b'GET /\r\n\r\n'
conn = h11.Connection(h11.SERVER)
conn.receive_data(request)
print(conn.next_event())
```

```py
import httptools

request = b'GET /\r\n\r\n'
class MyHandler:
    def on_url(self, url: bytes) -> None:
        print(url)
    def on_headers_complete(self) -> None:
        http_version = parser.get_http_version()
        method = parser.get_method()
        print(f"HTTP version: {http_version}, method: {method}")
parser = httptools.HttpRequestParser(MyHandler())
parser.feed_data(request)
```

**Comment by nggit** [comment]
Hi, something is missing in your request: 

> GET /<del> HTTP/1.0</del>\r\n\r\n


A minimally valid HTTP (1.0) request should look like this:
```
GET / HTTP/1.0\r\n\r\n
```

But in HTTP version 1.1 (the only one Uvicorn uses), `Host` [is required](https://datatracker.ietf.org/doc/html/rfc9112#name-changes-from-http-10):
```
GET / HTTP/1.1\r\nHost: localhost:8000\r\n\r\n
```

So don't forget to always add `Host` (and it should be in the first field), otherwise the server will likely respond to a `400 Bad Request`.
